Model/Sync/Checkout/Data.php

```diff
@@ -211,8 +211,9 @@ public function getDataBeforeChange()
      */
     public function prepareBillingAddress(Quote $quote)
     {
-        $billingAddress = $quote->getData('newBillingAddress') ?: $quote->getBillingAddress();
-        if (!$billingAddress->getCountryId() && $quote->getIsVirtual()) {
+        $newBillingAddress = $quote->getData('newBillingAddress');
```

Comment:
1. isVirtual && newBillingAddress -> newBillingAddress
2. isVirtual && !newBillingAddress -> defaultCustomerBillingAddress (because address details doesn't actually exist within the object).
3. !isVirtual && newBillingAddress -> newBillingAddress
4. !isVirtual && !newBillingAddress -> $quote->getBillingAddress() (becuase address details are being provided)